principles.rst

```diff
@@ -36,55 +36,273 @@ but lacks the unarticulated principles, which will be added over time.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Lexical Scoping Principle (LSP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Lexical Scoping Principle`:
+
+**Principle**:
+
+a. For every appearance of
+an identifier, it is possible to determine whether that appearance is a *binding site*
+or an *occurrence*, without involving the type system.
+
+b. For every *occurrence* of an
 identifier, it is possible to uniquely identify its *binding site*, without
 involving the type system.
 
-The LSP_ is true today, though Template Haskell splices may need to be run before
-completing name resolution (and running those splices requires type-checking them).
+The `Lexical Scoping Principle`_ is true today, with two complications:
+
+1. Template Haskell splices may need to be run before completing name resolution (and running those splices requires type-checking them).
+
+2. The `deprecated mechanism <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/duplicate_record_fields.html#selector-functions>`_ for disambiguating duplicate record fields violates the `Lexical Scoping Principle`_ by requiring the type system.
+
+*Motivation:* These principles mean that we can understand the binding
+structure of a program without relying on type inference, important both for the
+implementation of GHC and the sanity of programmers.
+
+(a) from `#448`_; (b) from `#378`_.
+
+Local Lexical Scoping Principle (LLSP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Local Lexical Scoping Principle`:
+
+**Principle**: For every appearance of an identifier, it is possible to determine
+whether that appearance is a *binding site* or an *occurrence*, without looking to see what identifiers are
+already in scope.
+
+This is a stronger version of the `Lexical Scoping Principle`_, part (a), that forbids even knowing what is in scope.
+
+The `Local Lexical Scoping Principle`_ is not true today, because of pattern signatures. If we have ``f (x :: Maybe a)``, the ``a``
+is an occurrence if ``a`` is already in scope, and it is a binding site otherwise.
+
+Along similar lines (though not strictly affected by the current phrasing of the `Local Lexical Scoping Principle`_), the
+type signature ``f :: a -> a`` might implicitly include ``forall a.`` or not, depending on whether ``a`` is in scope.
+
+*Motivation:* Tracking the set of in-scope variables is laborious for human readers. (The compiler is already
+doing this during name resolution.) This fact becomes even more poignant if we consider the possibility
+of mixing the term-level and type-level namespaces (`#270`_) and need to think about clashes between type
+variables and imported term variables.
+
+From `#448`_.
+
+Explicit Binding Principle (EBP)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _`Explicit Binding Principle`:
+
+**Principle**: Through the right combination of extensions and/or warning flags, it is possible
+for a Haskell programmer to ensure that all identifiers in a program have an explicit binding site.
```

Comment:
This one feels like a corollary of LLSP, does it need to be separate?

Reply:
LSP is about distinguishing occurrence from binding sites, EBP goes a step further and says that there must be a place for a (non-implicit) binding site. So slightly different. Haskell98 type variables are all occurrences (so LSP holds) but you can't explicitly bring them into scope (so no EBP). At least that's my reading.

Reply:
Yes, slightly different, but I think EBP follows directly from LSP, which is why I called it a corollary.

In other words, if

* We should be able to distinguish syntactically between bindings and occurrences, and
* Some bindings are invisible in Source Haskell

I think it follows rather directly that

* We should be able to explicitly write down binders that would be otherwise invisible

The other principles feel more orthogonal to each other than EBP vs LSP.

Reply:
See https://github.com/ghc-proposals/ghc-proposals/pull/448#issuecomment-1105771648 where I brought this up.

I agree even if we can lawyer out a slight difference --- do we actually *need* that difference? I our principles should feel simple and orthogonal, with lot of "elbow room" between them.

Reply:
Reading your comment, I think we're in agreement.

Reply:
I would definitely not say that this is a corollary, which would suggest that it's implied the LSP. You've articulated that something that could be implicit can also be explicit, which is far from obvious! (Indeed, much work in recent years is about making aspects of Haskell that have always been implicit more explicit.)

That said, I'm sympathetic to the desire for principles to be more orthogonal. Do you see a better refactoring?

Reply:
How about this?

* For every appearance of an identifier, determining whether it is a *binder* or *occurrence* should be a syntactic property.
    * This combines LSP (a) and LLSP
    * Other than TH this is true of Haskell today (see my response above about binders vs occurrences)

* For every occurrence of an identifier, determining its *binding site* should depend solely on the scoping rules (notably **not** the type system).
    * This is LSP (b)
    * What does it mean to determine the "binding site" of a variable? It means we can point to the **syntax** that bound the variable.

I claim that the fact that some binders cannot be written today in Source Haskell is a violation of these two principles. In the case of `f (x :: Maybe a) (y :: a) = ...`:

* Both `a`s are occurrences
* GHC inserts an implicit binder `f @a (x :: Maybe a) (y :: a) = ...` that binds `a`
* The implicit binder `@a` is not valid Source Haskell, so we cannot point to a piece of syntax that binds `a`. (If we could write the explicit binder ourselves I would call that close enough to satisfy the principles.)

Reply:
I like @gridaphobe's refactoring of the LSP and LLSP... but I still don't see how the EBP follows from what he has written. That is, what makes `f :: a -> a` OK but `f (x :: Maybe a)` not? I say the first is OK because we can write an explicit binder if we want to, whereas we can't for the second. But I don't see how this distinction arises from the principles @gridaphobe has written.

Reply:
It's not as crisp as I was hoping, but to me the problem is that if I can't write down an explicit binder, I can't point to the precise binding site. 

For example, what is the scope of the implicit `a` binder in `f (x :: Maybe a)`? Does it scope over just the current equation, or the entire definition of `f` like `f :: forall a. a -> a` would?

If we could say that `f (x :: Maybe a)` is just a shorthand for `f @a (x :: Maybe a)` it's much clearer that the scope is limited to the current equation.

---

That said, you mentioned earlier that there's a general pattern of allowing us to explicitly say things that were previously always implicit. I would also support encoding that as a broader principle, which would make this discussion moot. Typeclasses are a major exception to the explicitness principle, but there's certainly interest in allowing explicit selection and passing of dictionaries, we just don't have a great solution yet.

Reply:
> what makes `f :: a -> a` OK but `f (x :: Maybe a)` not?

Well, it's exactly that neither are 100% OK that makes me want `-XNo` to disable them. Two a first approximation we don't need to distinguish between them.

> but to me the problem is that if I can't write down an explicit binder, I can't point to the precise binding site.

Agreed. Well put.

Now that I am off my phone, I can write down the would-be desugaring in my head more explicitly as:

```haskel
f (x :: Maybe a) = ...
```
=>
```haskell
f = \ @(..) -> let a = _; in \ (x :: Maybe a) -> ...
```

> much clearer that the scope is limited to the current equation.

Indeed it's only due to the rather subtle fact that the definition being inferred is allowed to refer to all (explicit or inferred) binders that the scope is constrained.